nanocoat/lib/scritchui/scritchPencilRasterShapes.c

```diff
@@ -17,7 +17,399 @@
+	 * circle since less points will be sampled. The current value is a good
+	 * balance between filling all positions on all kinds of shapes while 
+	 * hitting as few pixels as possible.
+	 */
+
+	centerX = x + w / 2.0f;
+	centerY = y + h / 2.0f;
+	radiusX = w / 2.0f;
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
+	innerX, innerY;
+
+	firstFillX = lastFillX = firstFillY = lastFillY = -1;
+	
+	for (int i = 0; i < steps; i++) 
+	{
+		float angle = startAngleRad + (i * endAngleRad / steps);
+		
+		int innerX = round((centerX) + radiusX * cos(angle));
+		int innerY = round((centerY) + radiusY * sin(angle));
+		
+		/* Make sure we're not drawing out of bounds. */
+		if(innerX < clipLine->s.x || innerX >= clipLine->e.x
+			|| innerY < clipLine->s.y || innerY >= clipLine->e.y)
+			continue;
+
+		/* We cannot paint the same pixel more than once (breaks alpha) */
+		if((lastFillX == innerX && lastFillY == innerY) || 
+			(firstFillX == innerX && firstFillY == innerY)) 
+			continue;
+
+		if(i == 0) 
+		{ 
+			firstFillY = innerY; 
+			firstFillX = innerX; 
+		}
+		
+		lastFillX = innerX;
+		lastFillY = innerY;
+
+		/* If style is DOTTED, rendering will paint and skip pixels 1 by 1. */
+		if (dot)
+			error |= drawPixel(g, innerX, innerY);
+		dot ^= dotFlip;
+	}
+
+	/* Failed? */
+	if (sjme_error_is(error))
+		goto fail_any;
+		
+	/* Success? */
+	return error;
+	
+fail_any:
+	
+	return sjme_error_default(error);
+}
+
+sjme_errorCode sjme_scritchpen_corePrim_fillArc(
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint x,
+	sjme_attrInValue sjme_jint y,
+	sjme_attrInPositive sjme_jint w,
+	sjme_attrInPositive sjme_jint h,
+	sjme_attrInValue sjme_jint startAngle,
+	sjme_attrInValue sjme_jint arcAngle)
+{
+	sjme_errorCode error;
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	sjme_jint yz, yze, steps, innerX, innerY;
+	sjme_jboolean hasAlpha = 0;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
+		maxRad, angle;
+	sjme_jboolean *filledPixels = NULL;
+	sjme_scritchui_line* clipLine;
+	sjme_scritchui_pencilDrawPixelFunc drawPixel;
+
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+		
+	error = SJME_ERROR_NONE;
+
+	/* Java's coordinate system has positive angles moving counter-clockwise */
+	arcAngle = -arcAngle;
+	startAngle = -startAngle;
+
+	hasAlpha = g->hasAlpha;
+
+	drawPixel = g->prim.drawPixel;
+
+	/* Get clipping information. */
+	clipLine = &g->state.clipLine;
+
+	/** 
+	 * Only allocate the alpha buffer if the color isn't opaque. Noticeably
+	 * improves performance for opaque arcs.
+	 *
+	 * TODO: This can be optimized to take 1/8 of the usual memory size by
+	 * manipulating each bit of the boolean array below, as a _Bool variable
+	 * is often at least one whole byte in size.
```

Comment:
With `sjme_jint filledZ` and something like `(hasAlpha ? (filledPixels[filledZ >> 3] & (1 << (filledZ & 7))) : 1)` and `filledPixels[filledZ >> 3] |= (1 << (filledZ & 7))`, this can be reduced by 8 right now and would be preferrable.